Improve takeoffer error handing

ghubstan opened the pull request:
UI specific task runner validation errors cannot be passed to the client via a typical `ErrorMessageHandler`.  This change uses a new `GrpcErrorMessageHandler` implementation that places an optional `AvailabilityResult` in the `takeoffer` response if an offer is not available.  The client can derive a better failure reason message from the reply's AvailabilityResult.

- `GrpcErrorMessageHandler`  Add new `ErrorMessageHandler` implementation to get around the api specific problem of having to use an interface that is designed to build task error messages for the UI.
  
- `GrpcExceptionHandler`  Add a method for working with the  `ErrorMessageHandler` interface.

- `GrpcTradesService, CoreApi, CoreTradesService`:  Adjust the `takeoffer` error handling to give a failure reason provided by the new `GrpcErrorMessageHandler`. 

- `pb.proto`  Add `PRICE_CHECK_FAILED` to enum `AvailabilityResult`.

- `grpc.proto`  Add enum `AvailabilityResult` to `TakeOfferReply` message.  This field is populated by the `GrpcErrorMessageHandler` when `takeoffer` fails, making it easier to give CLI users a stylistically consistent failure reason.

- `OpenOfferManager`  Inject `CoreContex` to know isApiUser.  Set AvailabilityResult = `PRICE_CHECK_FAILED` if isApiUser=true, leaving it as `UNKNOWN_FAILURE` for isApiUser=false.

- OpenOfferManagerTest  Adjusted to new `CoreContext` constructor arg.  
  
- `GrpcClient`  Adjusted `takeOffer` to convert `AvailabilityResult` to user error message if the call fails -- does not return a trade.
  
- Added grpc exception catch blocks missing from some api test cases in `apitest/src/test/java/bisq/apitest/method/trade`.

ghubstan committed:
Adjust protos for takeoffer error handing

- Add PRICE_CHECK_FAILED to enum AvailabilityResult.
- Add enum AvailabilityResult to TakeOfferReply message.

ghubstan committed:
Inject CoreContext into OpenOfferManager, adjust test

ghubstan committed:
Adjust grpc & core services to new takeoffer error handling

- GrpcErrorMessageHandler  A new ErrorMessageHandler implementation
  to get around the api specific problem of having to use Task
  ErrorMessageHandlers that build task error messages for the UI.

- GrpcExceptionHandler  A new method for working with the
  ErrorMessageHandler interface.

- GrpcTradesService, CoreApi, CoreTradesService:  Ajdusted takeoffer
  error handling to give a failure reason provided by the new
  GrpcErrorMessageHandler.

ghubstan committed:
Convert AvailabilityResult to error msg when takeoffer fails

ghubstan committed:
Add missing grpc exception catch blocks

ghubstan commented:
PR https://github.com/bisq-network/bisq/pull/5292 should be reviewed and merged before this one.

ghubstan committed:
Derive AvailabilityResult from Task errorMessage

sqrrm commented on core/src/main/java/bisq/core/offer/AvailabilityResult.java:
```diff
@@ -23,11 +23,12 @@
     UNCONF_TX_LIMIT_HIT,
-    MAKER_DENIED_API_USER
+    MAKER_DENIED_API_USER,
+    PRICE_CHECK_FAILED
```
Just to confirm, this would fail to parse by old clients and thus use the first element in the enum, which is `UNKNOWN_FAILURE`, which makes sense.

ghubstan replied:
I confirmed this by sending `protobuf.AvailabilityResult.PRICE_CHECK_FAILED(13)` to a client that does not  have that proto enum value, and the client received `UNKNOWN_ENUM_VALUE_AvailabilityResult_13`.

wallclockbuilder reviewed (dismissed):
Tests passing.
LGTM.

ACK

ghubstan committed:
Merge branch 'master' into 01-fix-takeoffer-err-handling

ghubstan committed:
Refactor takeoffer reply proto & ErrorMessageHandler

- Add description msg TakeOfferReply proto, and fromProto method
  to core.offer.enum AvailabilityResult.  The description field
  maps a client usable error message to the enum.

- Adjust GrpcErrorMessageHandler to add AvailabilityResult.description()
  to takeoffer reply.

- Refactor (split up) GrpcClient's takeOffer.  Add getTakeOfferReply()
  to give clients a chance to make choices based on the reply's
  AvailabilityResult when the takeoffer command did not result in a
  trade. (Some errors are fatal, some not.)

sqrrm reviewed (approved):
utACK
